File: backend/routers/contact_lists.py
"""
contact_lists.py — Reusable audience lists for campaigns

GET    /api/contact-lists                        list all contact lists
POST   /api/contact-lists                        create list
GET    /api/contact-lists/{id}                   detail + members (paginated)
PUT    /api/contact-lists/{id}                   update name/description
DELETE /api/contact-lists/{id}                   delete list + members
POST   /api/contact-lists/{id}/members           add single member
POST   /api/contact-lists/{id}/upload            upload CSV (bulk add)
DELETE /api/contact-lists/{id}/members/{mid}     remove member
"""
import csv
import io
from datetime import datetime, timezone
from typing import Optional
from pydantic import BaseModel
from fastapi import APIRouter, Depends, HTTPException, Query, UploadFile, File
from sqlalchemy import text
from core.auth import require_pages
from core.database import get_db_pg

router = APIRouter()
ACCESS = require_pages(["campaigns"])


def _now():
    return datetime.now(timezone.utc)


class ListCreate(BaseModel):
    name:        str
    description: Optional[str] = None


class MemberCreate(BaseModel):
    first_name: Optional[str] = None
    last_name:  Optional[str] = None
    phone:      Optional[str] = None
    email:      Optional[str] = None
    cif_number: Optional[str] = None
    merge_data: dict          = {}


def _list_row(row) -> dict:
    return {
        "id":           row.id,
        "name":         row.name,
        "description":  row.description,
        "source":       row.source,
        "member_count": row.member_count,
        "created_by":   row.created_by,
        "created_at":   row.created_at.isoformat() if row.created_at else None,
        "updated_at":   row.updated_at.isoformat() if row.updated_at else None,
    }


def _member_row(row) -> dict:
    return {
        "id":         row.id,
        "list_id":    row.list_id,
        "first_name": row.first_name,
        "last_name":  row.last_name,
        "phone":      row.phone,
        "email":      row.email,
        "cif_number": row.cif_number,
        "merge_data": row.merge_data or {},
        "status":     row.status,
        "created_at": row.created_at.isoformat() if row.created_at else None,
    }


def _sync_count(db, list_id: int):
    row = db.execute(text(
        "SELECT COUNT(*) AS n FROM contact_list_members WHERE list_id=:id AND status='active'"
    ), {"id": list_id}).fetchone()
    db.execute(text("UPDATE contact_lists SET member_count=:n, updated_at=NOW() WHERE id=:id"),
               {"n": row.n if row else 0, "id": list_id})


# ── Lists ─────────────────────────────────────────────────────────────────────

@router.get("")
def list_contact_lists(
    limit: int = Query(100, ge=1, le=500),
    db = Depends(get_db_pg), user = Depends(ACCESS)
):
    rows = db.execute(text("""
        SELECT cl.*, u.full_name AS created_by_name
        FROM contact_lists cl
        LEFT JOIN o3c_users u ON cl.created_by=u.id
        ORDER BY cl.created_at DESC LIMIT :lim
    """), {"lim": limit}).fetchall()
    result = []
    for row in rows:
        d = _list_row(row)
        d["created_by_name"] = row.created_by_name
        result.append(d)
    return result


@router.post("", status_code=201)
def create_list(
    body: ListCreate, db = Depends(get_db_pg), user = Depends(ACCESS)
):
    row = db.execute(text("""
        INSERT INTO contact_lists (name, description, created_by)
        VALUES (:name, :desc, :by) RETURNING *
    """), {"name": body.name, "desc": body.description, "by": user["id"]}).fetchone()
    db.commit()
    return _list_row(row)


@router.get("/{list_id}")
def get_list(
    list_id: int,
    limit:  int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    search: Optional[str] = Query(None),
    db = Depends(get_db_pg), user = Depends(ACCESS)
):
    lst = db.execute(text("SELECT * FROM contact_lists WHERE id=:id"), {"id": list_id}).fetchone()
    if not lst:
        raise HTTPException(404, "List not found")

    where = "WHERE list_id=:id AND status='active'"
    params = {"id": list_id, "lim": limit, "off": offset}
    if search:
        where += " AND (first_name ILIKE :q OR last_name ILIKE :q OR phone ILIKE :q OR email ILIKE :q)"
        params["q"] = f"%{search}%"

    total = db.execute(text(f"SELECT COUNT(*) AS n FROM contact_list_members {where}"), params).fetchone()
    members = db.execute(text(f"""
        SELECT * FROM contact_list_members {where} ORDER BY id ASC LIMIT :lim OFFSET :off
    """), params).fetchall()

    d = _list_row(lst)
    d["total_members"] = total.n if total else 0
    d["members"]       = [_member_row(m) for m in members]
    return d


@router.put("/{list_id}")
def update_list(
    list_id: int, body: ListCreate,
    db = Depends(get_db_pg), user = Depends(ACCESS)
):
    if not db.execute(text("SELECT 1 FROM contact_lists WHERE id=:id"), {"id": list_id}).fetchone():
        raise HTTPException(404, "List not found")
    db.execute(text("""
        UPDATE contact_lists SET name=:name, description=:desc, updated_at=NOW() WHERE id=:id
    """), {"name": body.name, "desc": body.description, "id": list_id})
    db.commit()
    row = db.execute(text("SELECT * FROM contact_lists WHERE id=:id"), {"id": list_id}).fetchone()
    return _list_row(row)


@router.delete("/{list_id}", status_code=204)
def delete_list(
    list_id: int, db = Depends(get_db_pg), user = Depends(ACCESS)
):
    if not db.execute(text("SELECT 1 FROM contact_lists WHERE id=:id"), {"id": list_id}).fetchone():
        raise HTTPException(404, "List not found")
    db.execute(text("DELETE FROM contact_list_members WHERE list_id=:id"), {"id": list_id})
    db.execute(text("DELETE FROM contact_lists WHERE id=:id"), {"id": list_id})
    db.commit()


# ── Members ───────────────────────────────────────────────────────────────────

@router.post("/{list_id}/members", status_code=201)
def add_member(
    list_id: int, body: MemberCreate,
    db = Depends(get_db_pg), user = Depends(ACCESS)
):
    if not db.execute(text("SELECT 1 FROM contact_lists WHERE id=:id"), {"id": list_id}).fetchone():
        raise HTTPException(404, "List not found")
    import json as _json
    row = db.execute(text("""
        INSERT INTO contact_list_members
            (list_id, first_name, last_name, phone, email, cif_number, merge_data)
        VALUES (:lid, :fn, :ln, :phone, :email, :cif, :md::jsonb)
        RETURNING *
    """), {
        "lid":   list_id,
        "fn":    body.first_name,
        "ln":    body.last_name,
        "phone": body.phone,
        "email": body.email,
        "cif":   body.cif_number,
        "md":    _json.dumps(body.merge_data),
    }).fetchone()
    _sync_count(db, list_id)
    db.commit()
    return _member_row(row)


@router.post("/{list_id}/upload")
async def upload_csv(
    list_id: int,
    file: UploadFile = File(...),
    db = Depends(get_db_pg), user = Depends(ACCESS)
):
    """
    Upload a CSV to bulk-add contacts to a list.

    Required columns (at least one of phone/email):
      first_name, last_name, phone, email, cif_number

    Any extra columns are stored in merge_data (e.g. amount, due_date).
    """
    if not db.execute(text("SELECT 1 FROM contact_lists WHERE id=:id"), {"id": list_id}).fetchone():
        raise HTTPException(404, "List not found")
    if not file.filename.lower().endswith(".csv"):
        raise HTTPException(400, "File must be a CSV")

    import json as _json
    content = await file.read()
    text_content = content.decode("utf-8-sig", errors="replace")
    reader = csv.DictReader(io.StringIO(text_content))

    KNOWN = {"first_name","last_name","phone","email","cif_number"}
    inserted = 0
    errors   = []

    for i, row in enumerate(reader, 1):
        row = {k.strip().lower().replace(" ", "_"): (v.strip() if v else "") for k, v in row.items()}
        phone = row.get("phone", "") or None
        email = row.get("email", "") or None
        if not phone and not email:
            errors.append(f"Row {i}: no phone or email — skipped")
            continue
        merge = {k: v for k, v in row.items() if k not in KNOWN and v}
        try:
            db.execute(text("""
                INSERT INTO contact_list_members
                    (list_id, first_name, last_name, phone, email, cif_number, merge_data)
                VALUES (:lid, :fn, :ln, :phone, :email, :cif, :md::jsonb)
            """), {
                "lid":   list_id,
                "fn":    row.get("first_name") or None,
                "ln":    row.get("last_name")  or None,
                "phone": phone,
                "email": email,
                "cif":   row.get("cif_number") or None,
                "md":    _json.dumps(merge),
            })
            inserted += 1
        except Exception as e:
            errors.append(f"Row {i}: {str(e)[:80]}")

    _sync_count(db, list_id)
    db.execute(text(
        "UPDATE contact_lists SET source='csv', updated_at=NOW() WHERE id=:id"
    ), {"id": list_id})
    db.commit()
    return {"inserted": inserted, "errors": errors[:20]}


@router.delete("/{list_id}/members/{member_id}", status_code=204)
def remove_member(
    list_id: int, member_id: int,
    db = Depends(get_db_pg), user = Depends(ACCESS)
):
    if not db.execute(
        text("SELECT 1 FROM contact_list_members WHERE id=:id AND list_id=:lid"),
        {"id": member_id, "lid": list_id}
    ).fetchone():
        raise HTTPException(404, "Member not found")
    db.execute(text("DELETE FROM contact_list_members WHERE id=:id"), {"id": member_id})
    _sync_count(db, list_id)
    db.commit()

File: backend-go/handlers/contact_lists.go
package handlers

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"unicode"

	"github.com/go-chi/chi/v5"
	"github.com/o3c/reports/core"
)

func RegisterContactLists(r chi.Router, db *core.DB) {
	access := core.RequirePages("campaigns")
	r.With(access).Get("/", listContactLists(db))
	r.With(access).Post("/", createContactList(db))
	r.With(access).Get("/{id}", getContactList(db))
	r.With(access).Put("/{id}", updateContactList(db))
	r.With(access).Delete("/{id}", deleteContactList(db))
	r.With(access).Get("/{id}/members", listListMembers(db))
	r.With(access).Post("/{id}/members", addListMember(db))
	r.With(access).Put("/{id}/members/{mid}", updateListMember(db))
	r.With(access).Delete("/{id}/members/{mid}", removeListMember(db))
	r.With(access).Post("/{id}/preflight", preflightListCSV(db))
	r.With(access).Post("/{id}/upload", uploadListCSV(db))
}

func syncListCount(db *core.DB, r *http.Request, listID string) {
	if tr, _ := db.PGQuery(r.Context(),
		"SELECT COUNT(*) AS n FROM contact_list_members WHERE list_id=$1 AND status='active'", listID); len(tr) > 0 {
		db.PGExec(r.Context(), //nolint:errcheck
			"UPDATE contact_lists SET member_count=$1, updated_at=NOW() WHERE id=$2",
			toInt64(tr[0]["n"]), listID)
	}
}

func listContactLists(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		limit := qint(r, "limit", 100, 1, 500)
		offset := qint(r, "offset", 0, 0, 1<<30)
		from := r.URL.Query().Get("from")
		to := r.URL.Query().Get("to")

		where := "1=1"
		var filterArgs []any
		n := 1
		if from != "" {
			filterArgs = append(filterArgs, from)
			where += " AND cl.created_at::date >= $" + itoa(n) + "::date"
			n++
		}
		if to != "" {
			filterArgs = append(filterArgs, to)
			where += " AND cl.created_at::date <= $" + itoa(n) + "::date"
			n++
		}

		total := 0
		if tr, _ := db.PGQuery(r.Context(), "SELECT COUNT(*) AS n FROM contact_lists cl WHERE "+where, filterArgs...); len(tr) > 0 {
			total = int(toInt64(tr[0]["n"]))
		}
		args := append(append([]any(nil), filterArgs...), limit, offset)
		rows, err := db.PGQuery(r.Context(), fmt.Sprintf(`
			SELECT cl.*, u.full_name AS created_by_name
			FROM contact_lists cl
			LEFT JOIN o3c_users u ON cl.created_by=u.id
			WHERE %s ORDER BY cl.created_at DESC LIMIT $%d OFFSET $%d`, where, n, n+1), args...)
		if err != nil {
			respondErr(w, 500, "Query failed")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{ //nolint:errcheck
			"data":   rows,
			"total":  total,
			"limit":  limit,
			"offset": offset,
		})
	}
}

func createContactList(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var b struct {
			Name        string  `json:"name"`
			Description *string `json:"description"`
		}
		if err := json.NewDecoder(r.Body).Decode(&b); err != nil {
			respondErr(w, 400, "Invalid JSON")
			return
		}
		if b.Name == "" {
			respondErr(w, 422, "name is required")
			return
		}
		user := core.UserFromCtx(r.Context())
		rows, err := db.PGQuery(r.Context(),
			"INSERT INTO contact_lists (name, description, created_by) VALUES ($1,$2,$3) RETURNING *",
			b.Name, b.Description, user.ID)
		if err != nil {
			respondErr(w, 500, "Create failed")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(201)
		json.NewEncoder(w).Encode(rows[0]) //nolint:errcheck
	}
}

func getContactList(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		listRows, err := db.PGQuery(r.Context(), "SELECT * FROM contact_lists WHERE id=$1", id)
		if err != nil || len(listRows) == 0 {
			respondErr(w, 404, "List not found")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(listRows[0]) //nolint:errcheck
	}
}

func listListMembers(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		limit := qint(r, "limit", 200, 1, 1000)
		offset := qint(r, "offset", 0, 0, 1<<30)
		search := qstr(r, "q")

		if rows, _ := db.PGQuery(r.Context(), "SELECT 1 FROM contact_lists WHERE id=$1", id); len(rows) == 0 {
			respondErr(w, 404, "List not found")
			return
		}

		where := "list_id=$1 AND status='active'"
		args := []any{id}
		n := 2
		if search != "" {
			where += fmt.Sprintf(
				" AND (first_name ILIKE $%d OR last_name ILIKE $%d OR phone ILIKE $%d OR email ILIKE $%d OR cif_number ILIKE $%d)",
				n, n, n, n, n)
			args = append(args, "%"+search+"%")
			n++
		}
		filterArgs := append([]any(nil), args...)

		total := 0
		if tr, _ := db.PGQuery(r.Context(),
			fmt.Sprintf("SELECT COUNT(*) AS n FROM contact_list_members WHERE %s", where), filterArgs...); len(tr) > 0 {
			total = int(toInt64(tr[0]["n"]))
		}
		args = append(args, limit, offset)
		members, err := db.PGQuery(r.Context(),
			fmt.Sprintf("SELECT * FROM contact_list_members WHERE %s ORDER BY id ASC LIMIT $%d OFFSET $%d", where, n, n+1), args...)
		if err != nil {
			respondErr(w, 500, "Query failed")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{ //nolint:errcheck
			"data":   members,
			"total":  total,
			"limit":  limit,
			"offset": offset,
		})
	}
}

func updateContactList(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		var b struct {
			Name        string  `json:"name"`
			Description *string `json:"description"`
		}
		if err := json.NewDecoder(r.Body).Decode(&b); err != nil {
			respondErr(w, 400, "Invalid JSON")
			return
		}
		rows, err := db.PGQuery(r.Context(),
			"UPDATE contact_lists SET name=$1, description=$2, updated_at=NOW() WHERE id=$3 RETURNING *",
			b.Name, b.Description, id)
		if err != nil || len(rows) == 0 {
			respondErr(w, 404, "List not found")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(rows[0]) //nolint:errcheck
	}
}

func deleteContactList(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		guard, _ := db.PGQuery(r.Context(),
			"SELECT COUNT(*) AS n FROM campaigns WHERE list_id=$1", id)
		if len(guard) > 0 && toInt64(guard[0]["n"]) > 0 {
			respondErr(w, 409, "Cannot delete a contact list referenced by campaigns")
			return
		}
		db.PGExec(r.Context(), "DELETE FROM contact_list_members WHERE list_id=$1", id) //nolint:errcheck
		db.PGExec(r.Context(), "DELETE FROM contact_lists WHERE id=$1", id)             //nolint:errcheck
		w.WriteHeader(204)
	}
}

func addListMember(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		var b struct {
			FirstName *string        `json:"first_name"`
			LastName  *string        `json:"last_name"`
			Phone     *string        `json:"phone"`
			Email     *string        `json:"email"`
			CIFNumber *string        `json:"cif_number"`
			MergeData map[string]any `json:"merge_data"`
		}
		if err := json.NewDecoder(r.Body).Decode(&b); err != nil {
			respondErr(w, 400, "Invalid JSON")
			return
		}
		b.FirstName = cleanStringPtr(b.FirstName)
		b.LastName = cleanStringPtr(b.LastName)
		b.Phone = cleanStringPtr(b.Phone)
		b.Email = cleanStringPtr(b.Email)
		b.CIFNumber = cleanStringPtr(b.CIFNumber)
		if b.FirstName == nil && b.LastName == nil && b.Phone == nil && b.Email == nil && b.CIFNumber == nil {
			respondErr(w, 422, "at least one field is required (name, phone, email, or CIF number)")
			return
		}
		mergeJSON, _ := json.Marshal(b.MergeData)
		var phoneVal, emailVal string
		if b.Phone != nil {
			phoneVal = *b.Phone
		}
		if b.Email != nil {
			emailVal = *b.Email
		}
		rows, err := db.PGQuery(r.Context(), `
			INSERT INTO contact_list_members
			    (list_id, first_name, last_name, phone, email, phone_hmac, email_hmac, cif_number, merge_data)
			VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9::jsonb) RETURNING *`,
			id, b.FirstName, b.LastName, b.Phone, b.Email,
			nullStr(blindContactHMAC(phoneVal)), nullStr(blindContactHMAC(emailVal)),
			b.CIFNumber, string(mergeJSON))
		if err != nil {
			respondErr(w, 500, "Create failed")
			return
		}
		syncListCount(db, r, id)
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(201)
		json.NewEncoder(w).Encode(rows[0]) //nolint:errcheck
	}
}

func updateListMember(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		mid := chi.URLParam(r, "mid")
		var b struct {
			FirstName *string `json:"first_name"`
			LastName  *string `json:"last_name"`
			Phone     *string `json:"phone"`
			Email     *string `json:"email"`
			CIFNumber *string `json:"cif_number"`
		}
		if err := json.NewDecoder(r.Body).Decode(&b); err != nil {
			respondErr(w, 400, "Invalid JSON")
			return
		}
		b.FirstName = cleanStringPtr(b.FirstName)
		b.LastName = cleanStringPtr(b.LastName)
		b.Phone = cleanStringPtr(b.Phone)
		b.Email = cleanStringPtr(b.Email)
		b.CIFNumber = cleanStringPtr(b.CIFNumber)
		var phoneVal, emailVal string
		if b.Phone != nil {
			phoneVal = *b.Phone
		}
		if b.Email != nil {
			emailVal = *b.Email
		}
		rows, err := db.PGQuery(r.Context(), `
			UPDATE contact_list_members
			SET first_name=$1, last_name=$2, phone=$3, email=$4,
			    phone_hmac=$5, email_hmac=$6, cif_number=$7, updated_at=NOW()
			WHERE id=$8 AND list_id=$9 RETURNING *`,
			b.FirstName, b.LastName, b.Phone, b.Email,
			nullStr(blindContactHMAC(phoneVal)), nullStr(blindContactHMAC(emailVal)),
			b.CIFNumber, mid, id)
		if err != nil || len(rows) == 0 {
			respondErr(w, 404, "Member not found")
			return
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(rows[0]) //nolint:errcheck
	}
}

func removeListMember(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		mid := chi.URLParam(r, "mid")
		rows, _ := db.PGQuery(r.Context(),
			"SELECT 1 FROM contact_list_members WHERE id=$1 AND list_id=$2", mid, id)
		if len(rows) == 0 {
			respondErr(w, 404, "Member not found")
			return
		}
		db.PGExec(r.Context(), "DELETE FROM contact_list_members WHERE id=$1", mid) //nolint:errcheck
		syncListCount(db, r, id)
		w.WriteHeader(204)
	}
}

// parseContactCSV parses a CSV file into valid rows and validation errors.
// Used by both preflight and upload endpoints.
type csvContactRow struct {
	firstName *string
	lastName  *string
	phone     interface{}
	email     interface{}
	phoneHMAC *string
	emailHMAC *string
	cifNumber interface{}
	mergeJSON string
}

var knownContactCols = map[string]bool{
	"first_name": true, "last_name": true, "phone": true, "email": true, "cif_number": true,
}

func parseContactCSV(records [][]string) ([]csvContactRow, []string) {
	headers := make([]string, len(records[0]))
	for i, h := range records[0] {
		headers[i] = normaliseCSVHeader(h)
	}
	var valid []csvContactRow
	var errors []string
	for i, rec := range records[1:] {
		row := make(map[string]string, len(headers))
		for j, val := range rec {
			if j < len(headers) {
				row[headers[j]] = strings.TrimSpace(val)
			}
		}
		fn := strings.TrimSpace(row["first_name"])
		ln := strings.TrimSpace(row["last_name"])
		cif := strings.TrimSpace(row["cif_number"])
		phone := emptyToNil(row["phone"])
		email := emptyToNil(row["email"])
		if fn == "" && ln == "" && cif == "" && phone == nil && email == nil {
			errors = append(errors, fmt.Sprintf("Row %d: no identifiable fields — skipped", i+2))
			continue
		}
		merge := map[string]string{}
		for k, v := range row {
			if !knownContactCols[k] && v != "" {
				merge[k] = v
			}
		}
		mergeJSON, _ := json.Marshal(merge)
		valid = append(valid, csvContactRow{
			firstName: func() *string {
				if fn == "" {
					return nil
				}
				return &fn
			}(),
			lastName: func() *string {
				if ln == "" {
					return nil
				}
				return &ln
			}(),
			phone:     phone,
			email:     email,
			phoneHMAC: nullStr(blindContactHMAC(row["phone"])),
			emailHMAC: nullStr(blindContactHMAC(row["email"])),
			cifNumber: emptyToNil(cif),
			mergeJSON: string(mergeJSON),
		})
	}
	return valid, errors
}

func preflightListCSV(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if err := r.ParseMultipartForm(10 << 20); err != nil {
			respondErr(w, 400, "Cannot parse multipart form")
			return
		}
		file, header, err := r.FormFile("file")
		if err != nil {
			respondErr(w, 400, "file field required")
			return
		}
		defer file.Close()
		if !strings.HasSuffix(strings.ToLower(header.Filename), ".csv") {
			respondErr(w, 400, "File must be a CSV")
			return
		}
		records, err := csv.NewReader(file).ReadAll()
		if err != nil || len(records) < 2 {
			respondErr(w, 422, "Invalid CSV or empty file")
			return
		}
		valid, errors := parseContactCSV(records)
		maxErrors := 20
		if len(errors) < maxErrors {
			maxErrors = len(errors)
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{ //nolint:errcheck
			"total":   len(records) - 1,
			"valid":   len(valid),
			"invalid": len(errors),
			"errors":  errors[:maxErrors],
		})
	}
}

func uploadListCSV(db *core.DB) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id := chi.URLParam(r, "id")
		if err := r.ParseMultipartForm(10 << 20); err != nil {
			respondErr(w, 400, "Cannot parse multipart form")
			return
		}
		file, header, err := r.FormFile("file")
		if err != nil {
			respondErr(w, 400, "file field required")
			return
		}
		defer file.Close()
		if !strings.HasSuffix(strings.ToLower(header.Filename), ".csv") {
			respondErr(w, 400, "File must be a CSV")
			return
		}
		records, err := csv.NewReader(file).ReadAll()
		if err != nil || len(records) < 2 {
			respondErr(w, 422, "Invalid CSV or empty file")
			return
		}
		validRows, parseErrors := parseContactCSV(records)

		inserted := 0
		var insertErrors []string
		const batchSize = 500
		for start := 0; start < len(validRows); start += batchSize {
			end := start + batchSize
			if end > len(validRows) {
				end = len(validRows)
			}
			batch := validRows[start:end]
			var sb strings.Builder
			sb.WriteString("INSERT INTO contact_list_members (list_id, first_name, last_name, phone, email, phone_hmac, email_hmac, cif_number, merge_data) VALUES ")
			args := make([]interface{}, 0, len(batch)*9)
			for i, row := range batch {
				if i > 0 {
					sb.WriteString(",")
				}
				n := i*9 + 1
				fmt.Fprintf(&sb, "($%d,$%d,$%d,$%d,$%d,$%d,$%d,$%d,$%d::jsonb)", n, n+1, n+2, n+3, n+4, n+5, n+6, n+7, n+8)
				args = append(args, id, row.firstName, row.lastName, row.phone, row.email, row.phoneHMAC, row.emailHMAC, row.cifNumber, row.mergeJSON)
			}
			if _, err := db.PGExec(r.Context(), sb.String(), args...); err != nil {
				errMsg := err.Error()
				if len(errMsg) > 120 {
					errMsg = errMsg[:120]
				}
				insertErrors = append(insertErrors, fmt.Sprintf("Batch rows %d-%d: %s", start+2, end+1, errMsg))
			} else {
				inserted += len(batch)
			}
		}
		syncListCount(db, r, id)
		db.PGExec(r.Context(), "UPDATE contact_lists SET source='csv', updated_at=NOW() WHERE id=$1", id) //nolint:errcheck

		allErrors := append(parseErrors, insertErrors...)
		maxErrors := 20
		if len(allErrors) < maxErrors {
			maxErrors = len(allErrors)
		}
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(map[string]any{ //nolint:errcheck
			"inserted": inserted,
			"errors":   allErrors[:maxErrors],
		})
	}
}

// normaliseCSVHeader lowercases and snake_cases a CSV column header.
func normaliseCSVHeader(h string) string {
	h = strings.TrimSpace(h)
	var b strings.Builder
	for _, r := range h {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			b.WriteRune(unicode.ToLower(r))
		} else {
			b.WriteByte('_')
		}
	}
	return b.String()
}

func emptyToNil(s string) interface{} {
	if s == "" {
		return nil
	}
	return s
}

func cleanStringPtr(v *string) *string {
	if v == nil {
		return nil
	}
	s := strings.TrimSpace(*v)
	if s == "" {
		return nil
	}
	return &s
}
